internal/db/p2p_replicator.go

```diff
@@ -0,0 +1,341 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+	"encoding/json"
+
+	dsq "github.com/ipfs/go-datastore/query"
+
+	"github.com/sourcenetwork/corelog"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/event"
+	"github.com/sourcenetwork/defradb/internal/core"
+	"github.com/sourcenetwork/defradb/internal/merkle/clock"
+)
+
+func (db *db) SetReplicator(ctx context.Context, rep client.Replicator) error {
+	txn, err := db.NewTxn(ctx, false)
+	if err != nil {
+		return err
+	}
+	defer txn.Discard(ctx)
+
+	db.peerMutex.RLock()
+	if db.peerInfo.HasValue() && rep.Info.ID == db.peerInfo.Value().ID {
+		db.peerMutex.RUnlock()
+		return ErrSelfTargetForReplicator
+	}
+	db.peerMutex.RUnlock()
+	if err := rep.Info.ID.Validate(); err != nil {
+		return err
+	}
+
+	// TODO-ACP: Support ACP <> P2P - https://github.com/sourcenetwork/defradb/issues/2366
+	// ctx = db.SetContextIdentity(ctx, identity)
+	ctx = SetContextTxn(ctx, txn)
+
+	oldSchemas := make(map[string]struct{})
+	repKey := core.NewReplicatorKey(rep.Info.ID.String())
+	hasOldRep, err := txn.Systemstore().Has(ctx, repKey.ToDS())
+	if err != nil {
+		return err
+	}
+	if hasOldRep {
+		repBytes, err := txn.Systemstore().Get(ctx, repKey.ToDS())
+		if err != nil {
+			return err
+		}
+		oldRep := client.Replicator{}
+		err = json.Unmarshal(repBytes, &oldRep)
+		if err != nil {
+			return err
+		}
+		for _, schema := range oldRep.Schemas {
+			oldSchemas[schema] = struct{}{}
```

Comment:
question: Is this new code or old/moved code?

Reply:
is covered by integration test but now also covered by unit test